include/z64player.h

```diff
@@ -167,6 +167,78 @@ typedef enum PlayerItemAction {
     /* 0x53 */ PLAYER_IA_MAX
 } PlayerItemAction;
 
+#define GET_BSWORD_FROM_IA(itemAction) ((itemAction) - PLAYER_IA_SWORD_KOKIRI + 1)
+
+typedef enum PlayerBButtonSword {
+    /* 0 */ PLAYER_BSWORD_NONE,
+    /* 1 */ PLAYER_BSWORD_KOKIRI = GET_BSWORD_FROM_IA(PLAYER_IA_SWORD_KOKIRI),
+    /* 2 */ PLAYER_BSWORD_RAZOR = GET_BSWORD_FROM_IA(PLAYER_IA_SWORD_RAZOR),
+    /* 3 */ PLAYER_BSWORD_GILDED = GET_BSWORD_FROM_IA(PLAYER_IA_SWORD_GILDED),
+    /* 4 */ PLAYER_BSWORD_MAX
+} PlayerBButtonSword;
+
+#define GET_MELEE_WEAPON_FROM_IA(itemAction) ((itemAction) - PLAYER_IA_SWORD_KOKIRI + 1)
+
+typedef enum PlayerMeleeWeapon {
+    /* 0 */ PLAYER_MELEEWEAPON_NONE,
+    /* 1 */ PLAYER_MELEEWEAPON_SWORD_KOKIRI = GET_MELEE_WEAPON_FROM_IA(PLAYER_IA_SWORD_KOKIRI),
+    /* 2 */ PLAYER_MELEEWEAPON_SWORD_RAZOR = GET_MELEE_WEAPON_FROM_IA(PLAYER_IA_SWORD_RAZOR),
+    /* 3 */ PLAYER_MELEEWEAPON_SWORD_GILDED = GET_MELEE_WEAPON_FROM_IA(PLAYER_IA_SWORD_GILDED),
+    /* 4 */ PLAYER_MELEEWEAPON_SWORD_GREAT_FAIRY = GET_MELEE_WEAPON_FROM_IA(PLAYER_IA_SWORD_GREAT_FAIRY),
+    /* 5 */ PLAYER_MELEEWEAPON_STICK = GET_MELEE_WEAPON_FROM_IA(PLAYER_IA_STICK),
+    /* 6 */ PLAYER_MELEEWEAPON_ZORA_FINS = GET_MELEE_WEAPON_FROM_IA(PLAYER_IA_ZORA_FINS),
```

Comment:
Are these defined from the `itemAction` using a macro to help shiftability?

Reply:
Yes, and to help readability on the functions which map an IA to, say, a MELEEWEAPON
This is the function in master:
```c

s32 Player_ActionToMeleeWeapon(PlayerActionParam actionParam) {
    s32 weapon = actionParam - (PLAYER_AP_SWORD_KOKIRI - 1);

    if ((weapon > 0) && (weapon <= (PLAYER_AP_ZORA_FINS - (PLAYER_AP_SWORD_KOKIRI - 1)))) {
        return weapon;
    }
    return 0;
}
```
This is how it looks with this enum:
```c
PlayerMeleeWeapon Player_MeleeWeaponFromIA(PlayerItemAction itemAction) {
    PlayerMeleeWeapon weapon = GET_MELEE_WEAPON_FROM_IA(itemAction);

    if ((weapon > PLAYER_MELEEWEAPON_NONE) && (weapon < PLAYER_MELEEWEAPON_MAX)) {
        return weapon;
    }

    return PLAYER_MELEEWEAPON_NONE;
}
```

Another example is the bottle one:
```c

s32 Player_ActionToBottle(Player* player, PlayerActionParam actionParam) {
    s32 bottle = actionParam - PLAYER_AP_BOTTLE;

    // Relies on bottle-related action params to be contiguous
    if ((bottle >= (PLAYER_AP_BOTTLE - PLAYER_AP_BOTTLE)) && (bottle <= (PLAYER_AP_BOTTLE_FAIRY - PLAYER_AP_BOTTLE))) {
        return bottle;
    }

    return -1;
}

```

```c

PlayerBottle Player_BottleFromIA(Player* player, PlayerItemAction itemAction) {
    PlayerBottle bottle = GET_BOTTLE_FROM_IA(itemAction);

    if ((bottle > PLAYER_BOTTLE_NONE) && (bottle < PLAYER_BOTTLE_MAX)) {
        return bottle;
    }

    return PLAYER_BOTTLE_NONE;
}

```

Reply:
Oh, I think the existance of the enum is fine. I was more wondering about the usage of `GET_MELEE_WEAPON_FROM_IA` within the enum itself, since without it would be the same result. It feels cluttery, especially the Bottle one immediately below

Reply:
The point is to produce the same result in a matching build, and in case anybody modifies the `PLAYER_IA` enum (like adding a new sword after the great fairy one but before sticks) then the values would shift accordingly.

I'm not completely sure how I feel about using the macros here, and if people prefer to not use them (or remove those enums completely) then I can change it

Reply:
I think I see how it helps shiftability. But what if someone decided to add a 5th sword, such that:
```c
    /* 0x02 */ PLAYER_IA_SWORD_TRAINING,
    /* 0x03 */ PLAYER_IA_SWORD_KOKIRI,
    /* 0x04 */ PLAYER_IA_SWORD_RAZOR,
    /* 0x05 */ PLAYER_IA_SWORD_GILDED,
    /* 0x06 */ PLAYER_IA_SWORD_GREAT_FAIRY,
```
That would mean for:
`PLAYER_MELEEWEAPON_SWORD_TRAINING = GET_MELEE_WEAPON_FROM_IA(PLAYER_IA_SWORD_SWORD_TRAINING),`
that would evaluate to an enum value of 0, which would conflict with `/* 0 */ PLAYER_MELEEWEAPON_NONE`
I suppose it would work for all cases except for an enum 1 before the reference, i.e. `PLAYER_IA_SWORD_KOKIRI`. Thoughts?

Reply:
Or maybe just have a note saying `GET_MELEE_WEAPON_FROM_IA` must always be defined relative to the lowest value sword IA?

Reply:
An alternative would be to add "min values" for stuff like this
```c

    /* 0x03 */ PLAYER_IA_SWORD_MIN,
    /* 0x03 */ PLAYER_IA_SWORD_KOKIRI = PLAYER_IA_SWORD_MIN,
    /* 0x04 */ PLAYER_IA_SWORD_RAZOR,
    /* 0x05 */ PLAYER_IA_SWORD_GILDED,
    /* 0x06 */ PLAYER_IA_SWORD_GREAT_FAIRY,
```

and then define the macro like this
```c
#define GET_MELEE_WEAPON_FROM_IA(itemAction) ((itemAction) - PLAYER_IA_SWORD_MIN + 1)
```

This way it would be obvious what is needed to be changed in case somebody wanted to add a sword before `KOKIRI`

Reply:
Yeah, I think that would work

Reply:
Just pushed this change. Tell me what do you think
While doing this I noted the IA_BOTTLE was a bit weird as is, so I renamed it to `IA_BOTTLE_EMPTY`

Reply:
I do like `BOTTLE_EMPTY` but that can be applied to a lot of places I think. Not in this PR though, just something to note for later (using it for `IA` as you've already done is fine though, so no need to revert)